File: lld/problems/lru-cache/go/lru_cache_test.go
package lrucache

import (
	"sync"
	"testing"
)

func TestGetPutUpdateInPlace(t *testing.T) {
	c, err := NewLRUCache[string, int](2)
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}

	c.Put("a", 1)
	c.Put("b", 2)

	if v, ok := c.Get("a"); !ok || v != 1 {
		t.Fatalf("expected a=1, got %v ok=%v", v, ok)
	}

	c.Put("a", 100)
	if v, ok := c.Get("a"); !ok || v != 100 {
		t.Fatalf("expected a=100 after update, got %v ok=%v", v, ok)
	}
	if c.Len() != 2 {
		t.Fatalf("expected len 2, got %d", c.Len())
	}
}

func TestEvictsActualLRU(t *testing.T) {
	c, _ := NewLRUCache[string, int](2)
	c.Put("a", 1)
	c.Put("b", 2)

	// Touch "a" so "b" becomes the least-recently-used entry.
	c.Get("a")
	c.Put("c", 3)

	if _, ok := c.Get("b"); ok {
		t.Fatalf("expected b to be evicted")
	}
	if v, ok := c.Get("a"); !ok || v != 1 {
		t.Fatalf("expected a to survive eviction, got %v ok=%v", v, ok)
	}
	if v, ok := c.Get("c"); !ok || v != 3 {
		t.Fatalf("expected c to survive eviction, got %v ok=%v", v, ok)
	}
	if c.Len() != 2 {
		t.Fatalf("expected len 2, got %d", c.Len())
	}
}

func TestMissingKeyAndCapacityOne(t *testing.T) {
	c, _ := NewLRUCache[string, int](1)

	if _, ok := c.Get("missing"); ok {
		t.Fatalf("expected miss for absent key")
	}
	if c.Len() != 0 {
		t.Fatalf("expected len 0 after a miss, got %d", c.Len())
	}

	c.Put("a", 1)
	c.Put("b", 2)
	if _, ok := c.Get("a"); ok {
		t.Fatalf("expected a evicted at capacity 1")
	}
	if v, ok := c.Get("b"); !ok || v != 2 {
		t.Fatalf("expected b=2, got %v ok=%v", v, ok)
	}
}

// TestConcurrentAccess runs many goroutines performing puts/gets on a shared
// cache and asserts no panic occurs and the final size never exceeds capacity.
func TestConcurrentAccess(t *testing.T) {
	const capacity = 50
	const goroutines = 100

	c, _ := NewLRUCache[int, int](capacity)
	var wg sync.WaitGroup

	for i := 0; i < goroutines; i++ {
		wg.Add(1)
		go func(n int) {
			defer wg.Done()
			c.Put(n, n*n)
			c.Get(n)
			c.Get(n - 1)
		}(i)
	}
	wg.Wait()

	if c.Len() > capacity {
		t.Fatalf("expected size <= capacity %d, got %d", capacity, c.Len())
	}
	if len(c.items) != c.Len() {
		t.Fatalf("map size %d and reported len %d disagree", len(c.items), c.Len())
	}

	count := 0
	for n := c.head.next; n != c.tail; n = n.next {
		count++
	}
	if count != c.Len() {
		t.Fatalf("linked list length %d does not match cache len %d", count, c.Len())
	}
}

File: lld/problems/lru-cache/java/src/LRUCacheTest.java
import java.util.Optional;
import java.util.concurrent.CountDownLatch;

public class LRUCacheTest {

    public static void main(String[] args) {
        runAll();
    }

    public static void runAll() {
        testGetPutUpdateInPlace();
        testEvictsActualLRU();
        testMissingKeyAndCapacityOne();
        testConcurrentAccess();
        System.out.println("All LRUCacheTest cases passed.");
    }

    private static void testGetPutUpdateInPlace() {
        LRUCache<String, Integer> cache = new LRUCache<>(2);
        cache.put("a", 1);
        cache.put("b", 2);
        assertEquals(Optional.of(1), cache.get("a"), "get a");

        cache.put("a", 100);
        assertEquals(Optional.of(100), cache.get("a"), "get a after update");
        assertEquals(2, cache.size(), "size after update");
    }

    private static void testEvictsActualLRU() {
        LRUCache<String, Integer> cache = new LRUCache<>(2);
        cache.put("a", 1);
        cache.put("b", 2);
        cache.get("a");
        cache.put("c", 3);

        assertEquals(Optional.empty(), cache.get("b"), "b should be evicted");
        assertEquals(Optional.of(1), cache.get("a"), "a should survive");
        assertEquals(Optional.of(3), cache.get("c"), "c should survive");
        assertEquals(2, cache.size(), "size after eviction");
    }

    private static void testMissingKeyAndCapacityOne() {
        LRUCache<String, Integer> cache = new LRUCache<>(1);
        assertEquals(Optional.empty(), cache.get("missing"), "miss on empty cache");
        assertEquals(0, cache.size(), "size stays 0 after a miss");

        cache.put("a", 1);
        cache.put("b", 2);
        assertEquals(Optional.empty(), cache.get("a"), "a evicted at capacity 1");
        assertEquals(Optional.of(2), cache.get("b"), "b present at capacity 1");
    }

    private static void testConcurrentAccess() {
        final int capacity = 50;
        final int threads = 100;
        LRUCache<Integer, Integer> cache = new LRUCache<>(capacity);
        CountDownLatch latch = new CountDownLatch(threads);

        for (int i = 0; i < threads; i++) {
            final int n = i;
            new Thread(() -> {
                try {
                    cache.put(n, n * n);
                    cache.get(n);
                    cache.get(n - 1);
                } finally {
                    latch.countDown();
                }
            }).start();
        }
        try {
            latch.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        if (cache.size() > capacity) {
            throw new AssertionError("size " + cache.size() + " exceeds capacity " + capacity);
        }
        assertEquals(cache.size(), cache.listLength(), "map size and list length must agree");
    }

    private static void assertEquals(Object expected, Object actual, String label) {
        if (!expected.equals(actual)) {
            throw new AssertionError(label + ": expected " + expected + " but got " + actual);
        }
    }
}
